### PR TITLE
Allows users to send and accept/decline pal requests

### PR DESCRIPTION
**Pull request by anitacu**
ProfileServlet.java:
- 3 different forms: 
1) getParameter("bio"): sets the user bio (already done)
2) getParameter("accept")/getParameter("decline"): deals with accept/decline choice that user pressed [If accept, then add corresponding users to each other's list of pals. If either accept/decline, then remove requester from requestee's list of incoming requests and remove requestee from requester's list of outgoing requests.]
3) getParameter("palRequest"): works with a sent pal request. Adds requestee to list of requester's outgoing requests and adds requester to requestee's list of incoming requests.
After any of these forms, redirect user back to the page they are currently viewing.

User.java:
- modified addPal to only allow a user to add another name if they are not the same user [cannot be pals with oneself]
- added sentPalRequest(String name) to check if this user has sent some username a pal request.

profile.jsp:
1) if current user is viewing his/her own page, then can edit bio (already done) AND can view list of incoming requests to accept/decline any pal requests
2) if the current user and the person whose page they are viewing are pals, allow them to view to send a direct private message to this person 
Else if the current user and the person whose page they are viewing are not yet pals:
- if the current user has sent the person whose page they are viewing a pal request, it will state "Pal Request sent!"
- if the person whose page the current user is viewing has sent the current user a pal request, it will have the accept/decline buttons so they can respond
- else neither of these users have sent each other a pal request, so have a button to allow them to request each other as pals

** I wanted to keep this pull request focused on functionality and fairly simple/not too big, so I will include the tests for these different situations/conditions in the next pull request.

### CODE REVIEW

File: src/main/java/codeu/model/data/User.java [hunk 1]
```diff
@@ -141,9 +141,11 @@ public void setPals(List<String> pals) {
         return pals;
     }
 
-    /** Adds a pal to the list of this User's pals. */
+    /** Adds a pal to the list of this User's pals as long as the pal is not this User. */
     public void addPal(String name) {
```
Comment (tucosh) [inline src/main/java/codeu/model/data/User.java]: validate input?

Comment (tucosh) [reply src/main/java/codeu/model/data/User.java]: What if we've already added this user as a pal? Will s/he be a double pal?

Comment (anitacu) [reply src/main/java/codeu/model/data/User.java]: yes that is a good point! I added statements to check that the name passed in is a valid user (with only characters that were allowed in LoginServlet.java) and that the username is a registered user and that the name is not a pal already.

File: src/main/webapp/WEB-INF/view/profile.jsp [hunk 2]
```diff
@@ -4,6 +4,7 @@
 <%@ page import="java.util.List" %>
 <%@ page import="java.util.Iterator" %>
 <%@ page import="java.time.Instant" %>
+<%@ page import="java.util.ArrayList" %>
```
[the remark this answers is not in the document]
Comment (anitacu) [reply src/main/webapp/WEB-INF/view/profile.jsp]: whoops, just removed this unused import

File: src/main/java/codeu/controller/ProfileServlet.java [hunk 3]
```diff
@@ -116,22 +116,84 @@ public void doPost(HttpServletRequest request, HttpServletResponse response) thr
         String requestUrl = request.getRequestURI();
         String profileUsername = requestUrl.substring("/users/".length());
 
-        // gets user
-        User user = userStore.getUser(profileUsername);
+        // 3 forms: deal with updating bio, accept/decline pal request, or sending another user a pal request
+        if (request.getParameter("bio") != null) {
```
[the remark this answers is not in the document]
Comment (anitacu) [reply src/main/java/codeu/controller/ProfileServlet.java]: Hmm, @tucosh is it possible to have more than one doPost() method in one ProfileServlet class? How will profile.jsp passing the POST data know which of the 3 doPost methods to go to? maybe I'm not understanding this correctly?

[the remark this answers is not in the document]
Comment (tucosh) [reply src/main/java/codeu/controller/ProfileServlet.java]: No, I meant to shorten and simplify the doPost() method itself by moving all the disparate things it's doing here into separate methods (package private so they can be tested).

That is doPost() would look something like this:

if (request.getParameter("bio") != null) {
  handleBio(request);
} else if (something else) {
  doSomethingElse(request);
} else if(..) {
  do..(request);
}

.. and so would be much more readable and easier to understand at a glance. Each of the methods it call should be focused on a single task ("cohesive", remind me to discuss that at the next meeting..)

Is that clear?

[the remark this answers is not in the document]
Comment (anitacu) [reply src/main/java/codeu/controller/ProfileServlet.java]: yes thank you Henry! that makes a lot more sense :D

**Pull request merged**